### PR TITLE
Make Template actions comply  with phpreport DB abstractions

### PR DESCRIPTION
**Pull request by tonythomas01**
Needs more work with phpdoc and implementing the delete feature too.

### CODE REVIEW

**Comment by tonythomas01** [comment]
Need to convert the getTemplatesService to use the abstractions as a next step!

**Comment by tonythomas01** [comment]
This is now ready for review ;)

File: model/dao/TemplateDAO/PostgreSQLTemplateDAO.php [hunk 1]
```diff
@@ -0,0 +1,237 @@
+<?php
+/*
+ * Copyright (C) 2009-2013 Igalia, S.L. <info@igalia.com>
```
Comment (jaragunde) [inline model/dao/TemplateDAO/PostgreSQLTemplateDAO.php]: New file, new year :)

File: model/dao/TemplateDAO/PostgreSQLTemplateDAO.php [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+/** File for PostgreSQLTaskDAO
+ *
+ *  This file just contains {@link PostgreSQLTaskDAO}.
+ *
+ * @filesource
+ * @package PhpReport
+ * @subpackage DAO
+ * @author Jorge López Fernández <jlopez@igalia.com>
+ * @author Jacobo Aragunde Pérez <jaragunde@igalia.com>
```
Comment (jaragunde) [inline model/dao/TemplateDAO/PostgreSQLTemplateDAO.php]: We are not even using these `@author` tags lately, git information about authorship is more accurate (and gets updated for free!). Please, remove them :)

File: model/dao/TemplateDAO/PostgreSQLTemplateDAO.php [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+ */
+include_once(PHPREPORT_ROOT . '/util/DBPostgres.php');
+include_once(PHPREPORT_ROOT . '/model/vo/TemplateVO.php');
+include_once(PHPREPORT_ROOT . '/model/vo/UserVO.php');
+include_once(PHPREPORT_ROOT . '/model/vo/CustomerVO.php');
+include_once(PHPREPORT_ROOT . '/model/vo/ProjectVO.php');
+include_once(PHPREPORT_ROOT . '/model/dao/TemplateDAO/TemplateDAO.php');
+include_once(PHPREPORT_ROOT . '/util/ConfigurationParametersManager.php');
```
Comment (jaragunde) [inline model/dao/TemplateDAO/PostgreSQLTemplateDAO.php]: Check out these includes, you are not using most of them here.

File: model/dao/TemplateDAO/PostgreSQLTemplateDAO.php [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+        return $templateVO;
+    }
+
+    /** Template retriever by id for PostgreSQL.
+     *
+     * This function retrieves the row from Task table with the id <var>$taskId</var> and creates a {@link TaskVO} with its data.
+     *
+     * @param int $templateId the id of the row we want to retrieve.
+     * @return TemplateVO a value object {@link TaskVO} with its properties set to the values from the row.
```
Comment (jaragunde) [inline model/dao/TemplateDAO/PostgreSQLTemplateDAO.php]: Careful with the c&p, you are still talking about `TaskVO` objects here.

File: model/dao/TemplateDAO/PostgreSQLTemplateDAO.php [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+     */
+    public function getByUserId($userId) {
+        if (!is_numeric($userId))
+            throw new SQLIncorrectTypeException($userId);
+        $sql = "SELECT * FROM template WHERE usrid=$userId";
+        $result = $this->execute($sql);
+        return $result;
+    }
+
+    /** Task creator for PostgreSQL.
```
Comment (jaragunde) [inline model/dao/TemplateDAO/PostgreSQLTemplateDAO.php]: Same here :)

File: model/facade/TemplatesFacade.php [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+	/** Delete Templates function
+	 *
+	 * @param $templates
+	 * @throws null
+	 */
+	static function DeleteTemplates($templates) {
+		$action = new DeleteTemplateAction($templates);
+
+		$action->execute();
```
Comment (jaragunde) [inline model/facade/TemplatesFacade.php]: Watch out the indentation.

File: web/js/tasks.js [hunk 7]
```diff
@@ -590,6 +590,7 @@ var TaskPanel = Ext.extend(Ext.Panel, {
                     //create a new template record
                     var newTemplate = new templateRecord();
                     newTemplate.set('name', task.get('text'));
+                    newTemplate.set('text', task.get('text'));
```
Comment (jaragunde) [inline web/js/tasks.js]: Don't do that. As things are now, `task.get('text')' contains the template name, we are not interested in presenting that name in the description for tasks created from this template, users will always have to remove it.

File: web/services/createTemplatesService.php [hunk 8]
```diff
@@ -171,55 +175,29 @@
...
+                $string .= "<name>{$template->getName()}</name>";
+                $string .= "<story>{$template->getStory()}</story>";
+                $string .= "<telework>{$template->isTelework()}</telework>";
+                $string .= "<onsite>{$template->isOnsite()}</onsite>";
+                $string .= "<text>{$template->getText()}</text>";
+                $string .= "<ttype>{$template->getTtype()}</ttype>";
+                $string .= "<userId>{$template->getUserId()}</userId>";
+                $string .= "<projectId>{$template->getProjectId()}</projectId>";
+                $string .= "<customerId>{$template->getCustomerId()}</customerId>";
+                $string .= "<taskStoryId>{$template->getTaskStoryId()}</taskStoryId></template>";
```
Comment (jaragunde) [inline web/services/createTemplatesService.php]: I've just had an idea. Could you create a method TemplateVO->toXml() which would do this? You could also reuse it in `getUserTemplatesService.php`. We could extend it to existing VOs in the future.

File: web/services/getUserTemplatesService.php [hunk 9]
```diff
@@ -43,51 +44,41 @@
...
                 $string .= "true";
-            else
+            } else {
                 $string .= "false";
+            }
             $string .= "</onsite></template>";
         }
+        $string .= "</templates>";
     }
-    $string .= "</templates>";
```
Comment (jaragunde) [inline web/services/getUserTemplatesService.php]: The `<templates>` tag should be closed here, as it was before. If there were no stored templates, the service would return broken XML.

File: web/services/getUserTemplatesService.php [hunk 10]
```diff
@@ -22,6 +22,7 @@
 include_once(PHPREPORT_ROOT . '/web/services/WebServicesFunctions.php');
 include_once(PHPREPORT_ROOT . '/util/ConfigurationParametersManager.php');
 include_once(PHPREPORT_ROOT . '/util/DBPostgres.php');
```
Comment (jaragunde) [inline web/services/getUserTemplatesService.php]: You probably don't need the two includes above, please double-check the three services :)

**Comment by jaragunde** [comment]
Regarding #265, the description of the task (the "text" field) is not yet being saved. But the UI to get the template name looks really cool!

One more thing, could you update the [user documentation](https://github.com/Igalia/phpreport/blob/master/docs/user/tasks.rst#using-templates) with the new behavior of the templates?

I think we could merge the entire feature as soon as you get that couple of things done, and we would close two tickets!

**Comment by tonythomas01** [comment]
The updation of the documentation seems to be to the master branch. Should I just wait for this one to be 1:1 to master, and then do a PR to `master` ?

**Comment by jaragunde** [comment]
You could do the documentation changes in this PR too; they haven't been updated recently.

**Pull request merged**